include/glow/Backends/ExecutionContext.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_BACKENDS_EXECUTIONCONTEXT_H
+#define GLOW_BACKENDS_EXECUTIONCONTEXT_H
+
+#include "glow/Backends/TraceEvents.h"
+#include "glow/Graph/PlaceholderBindings.h"
+
+#include "llvm/ADT/STLExtras.h"
+
+namespace glow {
+
+/// Sub-classed per backend, this holds Device specific per-function information
+/// if that is necessary on that particular backend.
+class DeviceBindings {
+public:
+  DeviceBindings() = default;
+  virtual ~DeviceBindings() {}
+
+  // Base class has no members.
+
+  virtual std::unique_ptr<DeviceBindings> clone() {
+    return llvm::make_unique<DeviceBindings>();
+  }
+};
+
+/// The runtime context for a single execution (Inferance or Training) in the
+/// the Glow Execution Engine or HostManager. This class includes the mapping
+/// between Input/Output Placeholders and the materialized Tensors used for this
+/// run, the set of Device specific details required to execute the function,
+/// and stores TraceEvents that were generated as a result of the run.
+class ExecutionContext {
+  std::unique_ptr<PlaceholderBindings> placeholderBindings_;
+  std::unique_ptr<DeviceBindings> deviceBindings_;
+
+  /// Trace Events recorded during this run.
+  std::vector<TraceEvent> traceEvents_;
+
+public:
+  ExecutionContext()
+      : placeholderBindings_(llvm::make_unique<PlaceholderBindings>()),
+        deviceBindings_(llvm::make_unique<DeviceBindings>()) {}
```

Comment:
I think deviceBindings can be a nullptr if one is not provided. The base class provides no value to a compiledFunction so it could just check for a nullptr.